File: impl/src/org/ufomatics/archipelago/control/UFOPlatformControllerImpl.java
package org.ufomatics.archipelago.control;

import org.ufomatics.archipelago.model.*;
import org.ufomatics.archipelago.util.Localizer;
import sun.misc.Launcher;

import java.beans.XMLDecoder;
import java.beans.XMLEncoder;
import java.io.*;
import java.lang.reflect.Modifier;
import java.net.URL;
import java.util.HashSet;
import java.util.Properties;
import java.util.Set;
import java.util.StringTokenizer;
import java.util.prefs.BackingStoreException;
import java.util.prefs.Preferences;

/**
 * @author Jérôme Beau
 * @version 7 mai 2006 16:57:19
 */
public class UFOPlatformControllerImpl implements UFOPlatformController {
    private Localizer localizer;
    private Set<DatabaseAdapter> adapters = new HashSet<DatabaseAdapter>();
    private MetaMapping metaMapping = new MetaMappingImpl();
    private Preferences preferences = Preferences.userNodeForPackage(getClass());

    public UFOPlatformControllerImpl(Localizer localizer) {
        this.localizer = localizer;
        initAdapters();
        initMetaMapping();
    }

    private void initMetaMapping() {
        metaMapping = new MetaMappingImpl();
    }

    public MetaModel getMetaModel() {
        return metaMapping.getMetaModel();
    }

    public boolean hasMetaModel() {
        return metaMapping.getMetaModel() != null && !metaMapping.getMetaModel().isEmpty();
    }

    public void loadMetaMapping(String name) throws MetaException {
        System.out.println("Loading " + name);
        try {
            FileInputStream fileInputStream = new FileInputStream(name);
            InputStream inputStream = new BufferedInputStream(fileInputStream);
            XMLDecoder xmlDecoder = new XMLDecoder(inputStream);
            try {
                Object o = xmlDecoder.readObject();
                if (o == null) {
                    throw new MetaException("Could not resolve content of file " + name);
                }
                if (o instanceof MetaModel) {
                    metaMapping.setMetaModel((MetaModel) o);
                } else {
                    metaMapping = (MetaMapping) o;
                }
                Set<String> recentFiles = getRecentFiles();
                recentFiles.add(name);
                try {
                    preferences.flush();
                } catch (BackingStoreException e) {
                    e.printStackTrace();
                }
            } finally {
                xmlDecoder.close();
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            throw new MetaException("Could not find file " + name, e);
        }
    }

    public Set<String> getRecentFiles() {
        Set<String> recentFiles = new HashSet<String>();
        String recentFilesStr = preferences.get("recentFiles", "");
        StringTokenizer st = new StringTokenizer(recentFilesStr, ";");
        while (st.hasMoreTokens()) {
            String s = st.nextToken();
            recentFiles.add(s);
        }
        return recentFiles;
    }

    public void saveMetaMapping(String filename) throws IOException {
        OutputStream outputStream = new BufferedOutputStream(new FileOutputStream(filename));
        XMLEncoder xmlEncoder = new XMLEncoder(outputStream);
        try {
            xmlEncoder.writeObject(metaMapping);
        } finally {
            xmlEncoder.close();
        }
    }

    public MetaType getDefaultFieldType() {
        return MetaTypeImpl.TEXT;
    }

    public DatabaseAdapter getDefaultAdapter() {
        return adapters.isEmpty() ? null : adapters.iterator().next();
    }

    private void initAdapters() {
        Set<Class> adapterClasses = find(DatabaseAdapter.class);
        Properties properties = new Properties();
        for (Class adapterClass : adapterClasses) {
            try {
                DatabaseAdapter adapter = (DatabaseAdapter) adapterClass.newInstance();
                adapter.setProperties(properties);
                adapters.add(adapter);
            } catch (InstantiationException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        }
    }

    public static Set<Class> find(Class tosubclass) {
        Set<Class> classes = new HashSet<Class>();
        Package[] pcks = Package.getPackages();
        for (Package pck : pcks) {
            classes.addAll(find(pck.getName(), tosubclass));
        }
        return classes;
    }

    public static Set<Class> find(String pckgname, Class toSubClass) {
        Set<Class> classes = new HashSet<Class>();
        String name = new String(pckgname);
        if (!name.startsWith("/")) {
            name = "/" + name;
        }

        name = name.replace('.', '/');

        // Get a File object for the package
        URL url = Launcher.class.getResource(name);
        if (url != null) {
            File directory = new File(url.getFile());

            if (directory.exists()) {
                // Get the list of the files contained in the package
                String[] files = directory.list();
                for (String file : files) {
                    // we are only interested in .class files
                    if (file.endsWith(".class")) {
                        // removes the .class extension
                        String classname = file.substring(0, file.length() - 6);
                        try {
                            // Try to create an instance of the object
                            Class aClass = Class.forName(pckgname + "." + classname);
                            if (((aClass.getModifiers() & Modifier.ABSTRACT) == 0) && toSubClass.isAssignableFrom(aClass)) {
                                classes.add(aClass);
                            }
                        } catch (ClassNotFoundException cnfex) {
                            System.err.println(cnfex);
                        }
                    }
                }
            }
        }
        return classes;
    }

    public Localizer getLocalizer() {
        return localizer;
    }

    public Set<DatabaseAdapter> getAdapters() {
        return adapters;
    }

    public void close() {
        for (DatabaseAdapter adapter : adapters) {
            adapter.close();
        }
    }

    public Set<Database> getDatasources() {
        return metaMapping.getDatasources();
    }

    public void addDatasource(Database database) {
        metaMapping.getDatasources().add(database);
    }

    public Database createDatasource() {
        return new DatabaseImpl();
    }
}

File: src/archipelago/impl/src/ufomatics/archipelago/control/UFOPlatformControllerImpl.ts
import {DatabaseAdapter} from "archipelago/api/model/DatabaseAdapter"
import {Localizer} from "archipelago/api/util/Localizer"
import {MetaMapping} from "archipelago/api/model/MetaMapping"
import {MetaMappingImpl} from "archipelago/impl/src/ufomatics/archipelago/model/MetaMappingImpl"
import {MetaType, TEXT} from "archipelago/api/model/MetaType"
import {Properties} from "archipelago/api/util/jsdk/util/Properties"
import {Database} from "archipelago/api/model/Database"
import {DatabaseImpl} from "archipelago/impl/src/ufomatics/archipelago/model/DatabaseImpl"
import {UFOPlatformController} from "archipelago/api/control/UFOPlatformController"
import {MetaException} from "archipelago/api/model/MetaException"
import {MetaModel} from "archipelago/api/model/MetaModel"
import {HashSet} from "archipelago/api/util/jsdk/util/HashSet"
import {StringTokenizer} from "archipelago/api/util/jsdk/util/StringTokenizer"

/**
 *
 */
export class UFOPlatformControllerImpl implements UFOPlatformController {
  private adapters = new HashSet<DatabaseAdapter>()
  private metaMapping: MetaMapping = new MetaMappingImpl()
  private preferences = Preferences.userNodeForPackage(getClass())

  constructor(protected localizer: Localizer) {
    this.initAdapters()
    this.initMetaMapping()
  }

  static find(tosubclass: Class): Set<Class> {
    const classes = new Set<Class>()
    const pcks = Package.getPackages()
    for (let pck of pcks) {
      classes.addAll(find(pck.getName(), tosubclass))
    }
    return classes
  }

  static find(pckgname: string, toSubClass: Class): Set<Class> {
    const classes = new Set<Class>()
    const name = String(pckgname)
    if (!name.startsWith("/")) {
      name = "/" + name
    }
    name = name.replace(".", "/")

    // Get a File object for the package
    const url = Launcher.class.getResource(name)
    if (url != null) {
      const directory = new File(url.getFile())
      if (directory.exists()) {
        // Get the list of the files contained in the package
        const files = directory.list()
        for (let file of files) {
          // we are only interested in .class files
          if (file.endsWith(".class")) {
            // removes the .class extension
            const classname = file.substring(0, file.length() - 6)
            try {
              // Try to create an instance of the object
              const aClass = Class.forName(pckgname + "." + classname)
              if (((aClass.getModifiers() & Modifier.ABSTRACT) == 0) && toSubClass.isAssignableFrom(aClass)) {
                classes.add(aClass)
              }
            } catch (cnfex) {
              console.error(cnfex)
            }
          }
        }
      }
    }
    return classes
  }

  getMetaModel(): MetaModel {
    return this.metaMapping.getMetaModel()
  }

  hasMetaModel(): boolean {
    return this.metaMapping.getMetaModel() != null && !this.metaMapping.getMetaModel().isEmpty()
  }

  loadMetaMapping(name: string): void {
    console.log("Loading " + name)
    try {
      const fileInputStream = new FileInputStream(name)
      const inputStream = new BufferedInputStream(fileInputStream)
      const xmlDecoder = new XMLDecoder(inputStream)
      try {
        const o = xmlDecoder.readObject()
        if (o == null) {
          throw new MetaException("Could not resolve content of file " + name)
        }
        if (o instanceof MetaModel) {
          this.metaMapping.setMetaModel(o as MetaModel)
        } else {
          this.metaMapping = o as MetaMapping
        }
        const recentFiles = this.getRecentFiles()
        recentFiles.add(name)
        try {
          this.preferences.flush()
        } catch (e) {
          e.printStackTrace()
        }
      } finally {
        xmlDecoder.close()
      }
    } catch (e) {
      console.error(e)
      throw new MetaException("Could not find file " + name, e)
    }
  }

  getRecentFiles(): HashSet<String> {
    const recentFiles = new HashSet<String>()
    const recentFilesStr = this.preferences.get("recentFiles", "")
    const st = new StringTokenizer(recentFilesStr, ";")
    while (st.hasMoreTokens()) {
      const s = st.nextToken()
      recentFiles.add(s)
    }
    return recentFiles
  }

  saveMetaMapping(filename: string): void {
    const outputStream = new BufferedOutputStream(new FileOutputStream(filename))
    const xmlEncoder = new XMLEncoder(outputStream)
    try {
      xmlEncoder.writeObject(this.metaMapping)
    } finally {
      xmlEncoder.close()
    }
  }

  getDefaultFieldType(): MetaType {
    return TEXT
  }

  getDefaultAdapter(): DatabaseAdapter {
    return this.adapters.isEmpty() ? null : this.adapters.values().next()
  }

  getLocalizer(): Localizer {
    return this.localizer
  }

  getAdapters(): HashSet<DatabaseAdapter> {
    return this.adapters
  }

  close(): void {
    for (let adapter of this.adapters._set) {
      adapter.close()
    }
  }

  getDatasources(): Set<Database> {
    return this.metaMapping.getDatasources()
  }

  addDatasource(database: Database): void {
    this.metaMapping.getDatasources().add(database)
  }

  createDatasource(): Database {
    return new DatabaseImpl()
  }

  private initMetaMapping(): void {
    this.metaMapping = new MetaMappingImpl()
  }

  private initAdapters(): void {
    const adapterClasses = UFOPlatformControllerImpl.find(DatabaseAdapter.class)
    const properties = new Properties()
    for (let adapterClass of adapterClasses) {
      try {
        const adapter = adapterClass.newInstance() as DatabaseAdapter
        adapter.setProperties(properties)
        this.adapters.add(adapter)
      } catch (e) {
        console.error(e)
      }
    }
  }
}
